### PR TITLE
Switch to eqx

### PR DESCRIPTION
**Pull request by hvater**
Migrated from [jax_dataclasses ](https://github.com/brentyi/jax_dataclasses) to [equinox ](https://docs.kidger.site/equinox/)for states etc.

Additionally, cleaned up the vmapping procedure

### CODE REVIEW

File: exciting_environments/core_env.py [hunk 1]
```diff
@@ -1,18 +1,25 @@
...
+from jax.tree_util import tree_flatten, tree_unflatten, tree_structure, tree_leaves
+import equinox as eqx
 import diffrax
 import chex
 
 
-class CoreEnvironment(ABC):
+class CoreEnvironment(eqx.Module):
+    tau: float = eqx.field(static=True)
+    _solver: diffrax.AbstractSolver = eqx.field(static=True)
```
Comment (hvater) [inline exciting_environments/core_env.py:18]: I am not sure if this should be `static`, as it is a general type and could theoretically have `jax.Array`s inside which in turn would make it immutable. 
See comments in https://docs.kidger.site/equinox/api/module/advanced_fields/#equinox.field

In the end, it is probably fine for our purposes, but we should keep an eye on this.
Does it make any difference to make this not `static`?

File: exciting_environments/core_env.py [hunk 2]
```diff
@@ -107,30 +106,28 @@ class Action:
 
     @partial(jax.jit, static_argnums=0)
```
Comment (hvater) [inline exciting_environments/core_env.py]:
```suggestion
    @eqx.filter_jit
```

Also for all other cases: Any reason not to do this?

File: exciting_environments/core_env.py [hunk 3]
```diff
@@ -107,30 +106,28 @@ class Action:
...
             action: The action applied at the current step.
-            static_params: Static parameters of the environment.
 
         Returns:
             state: The updated state after one simulation step.
         """
         return
 
-    @partial(jax.jit, static_argnums=[0, 4, 5])
+    @partial(jax.jit, static_argnums=[0, 3, 4])
```
Comment (hvater) [inline exciting_environments/core_env.py]:
```suggestion
    @eqx.filter_jit
```

And so on and so forth...

File: exciting_environments/core_env.py [hunk 4]
```diff
@@ -107,30 +106,28 @@ class Action:
...
     @abstractmethod
-    def _ode_solver_simulate_ahead(self, init_state, actions, static_params, obs_stepsize, action_stepsize):
+    def _ode_solver_simulate_ahead(self, init_state, actions, obs_stepsize, action_stepsize):
```
Comment (hvater) [inline exciting_environments/core_env.py:124]:
```suggestion
    def _ode_solver_simulate_ahead(self, init_state, actions, obs_stepsize=None, action_stepsize=None):
    
    if obs_stepsize=None:
        obs_stepsize = self.tau
        
    if action_stepsize=None:
        action_stepsize= self.tau
```
Any reason not to do this? Do you have any idea?

Comment (hvater) [reply exciting_environments/core_env.py:124]: I saw that you do this elsewhere. A good change :)

File: exciting_environments/pmsm/motor_parameters.py [hunk 5]
```diff
@@ -6,146 +6,198 @@
...
-                name,
-                jax.nn.relu(jnp.abs(getattr(physical_state_norm, name)) - 1.0),
-            )
+    phys_soft_const = jax.tree.map(lambda x: jax.nn.relu(jnp.abs(x) - 1.0), physical_state_norm)
     return phys_soft_const, None
 
 
+brusa_lut_raw = loadmat(Path(__file__).parent / Path("LUT_BRUSA_jax_grad.mat"))
+
+brusa_interpolators, brusa_lut = generate_interpolators_and_lut(brusa_lut_raw)
```
Comment (hvater) [inline exciting_environments/pmsm/motor_parameters.py]: Two things:

- I would like these to be uppercase if they are global variables, i.e.
```suggestion
BRUSA_LUT_RAW = loadmat(Path(__file__).parent / Path("LUT_BRUSA_jax_grad.mat"))

BRUSA_INTERPOLATORS, BRUSA_LUT = generate_interpolators_and_lut(brusa_lut_raw)
```
- I am not sure how dangerous it is that you accidentally change these inplace to some other values.'
Maybe it would be preferable to have a function that loads the LUTs on request, rather than with the package initialization. What do you think?

File: exciting_environments/pmsm/motor_parameters.py [hunk 6]
```diff
@@ -6,146 +6,198 @@
...
         deadtime=1,
     ),
     default_soft_constraints=default_soft_constraints,
-    pmsm_lut=loadmat(Path(__file__).parent / Path("LUT_BRUSA_jax_grad.mat")),
+    interpolators=brusa_interpolators,
 )
 
+sew_lut_raw = loadmat(Path(__file__).parent / Path("LUT_SEW_jax_grad.mat"))
+
+sew_interpolators, sew_lut = generate_interpolators_and_lut(sew_lut_raw)
```
Comment (hvater) [inline exciting_environments/pmsm/motor_parameters.py]: See comment on BRUSA LUT and interpolators

File: exciting_environments/core_env.py [hunk 7]
```diff
@@ -219,213 +211,186 @@ def generate_truncated(self, state, env_properties):
...
         return
 
-    @jdc.pytree_dataclass
-    class State:
+    class State(eqx.Module):
         """The state of the environment."""
 
-        physical_state: jdc.pytree_dataclass
+        physical_state: eqx.Module
         PRNGKey: jax.Array
```
Comment (hvater) [inline exciting_environments/core_env.py]:
```suggestion
        prng_key: jax.Array
```
I just realized that we are using CamelCase here, which is reserved for class definitions. Please change this for all occurences in the repository

File: exciting_environments/pmsm/pmsm_env.py [hunk 8]
```diff
@@ -113,9 +114,12 @@ def clip_in_abc_coordinates(u_dq, u_dc, omega_el, eps, tau):
 
 
 class PMSM(CoreEnvironment):
+    control_state: list = eqx.field(static=True)
+    soft_constraints_logic: Callable = eqx.field(static=True)
+    LUT_interpolators: dict
```
Comment (hvater) [inline exciting_environments/pmsm/pmsm_env.py]: _Non-blocking_
Maybe the LUTs should also be a class object and not a dict. It is not important. Just an idea

File: exciting_environments/pmsm/pmsm_env.py [hunk 9]
```diff
@@ -276,10 +253,9 @@ class StaticParams:
         l_q: jax.Array
         psi_p: jax.Array
         u_dc: jax.Array
-        deadtime: jax.Array
+        deadtime: jax.Array = eqx.field(static=True)
```
Comment (hvater) [inline exciting_environments/pmsm/pmsm_env.py]: `jax.Array`s should not be a static field unless we have a very good reason. Is this a jax array because of vmapping? In any case, I think the deadtime should be about as static as the other quantities in the static parames, so it should be fine to remove it

```suggestion
        deadtime: jax.Array
```

File: exciting_environments/pendulum/pendulum_env.py [hunk 10]
```diff
@@ -401,3 +397,381 @@ def obs_description(self):
     @property
     def action_description(self):
         return np.array(["torque"])
+
```
Comment (hvater) [inline exciting_environments/pendulum/pendulum_env.py]: No reason to keep this. Just remove all that you have commented out here.

File: examples/env_functions_examples.ipynb [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+    "import gymnasium as gym\n",
+    "import exciting_environments as excenv\n",
+    "#from exciting_environments import GymWrapper, EnvironmentRegistry\n",
+    "#import jax_dataclasses as jdc\n",
+    "from exciting_environments import EnvironmentRegistry\n",
+    "from dataclasses import fields\n",
+    "from exciting_environments.utils import MinMaxNormalization\n",
+    "jax.config.update(\"jax_enable_x64\", True)\n",
+    "import diffrax\n",
+    "from exciting_environments.pmsm.motor_parameters import MotorVariant"
```
Comment (hvater) [inline examples/env_functions_examples.ipynb]:
```suggestion
    "import jax.numpy as jnp\n"
    "import jax\n"
    "from exciting_environments import EnvironmentRegistry\n"
    "jax.config.update(\"jax_enable_x64\", True)\n"
```

Not very important, but there is a lot of unused imports here

File: examples/env_functions_examples.ipynb [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+   "outputs": [],
+   "source": [
+    "key=jax.random.PRNGKey(1234)\n",
+    "obs, state = pend_env.reset()\n",
+    "generated_observations = []\n",
+    "generated_actions= []\n",
+    "generated_observations.append(obs)\n",
+    "for i in range(4):\n",
+    "    key,subkey= jax.random.split(key)\n",
+    "    action = jnp.ones(2)#jax.random.uniform(subkey,(2,),minval=-1,maxval=1)\n",
```
Comment (hvater) [inline examples/env_functions_examples.ipynb:81]: Why are the actions 2-dim for the pendulum and why is the API fine with it? Is this intended?

File: examples/env_functions_examples.ipynb [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cdfab95d",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "batch_size= 2\n",
+    "keys = jax.random.split(jax.random.PRNGKey(0), batch_size)\n",
+    "envs = [EnvironmentRegistry.PENDULUM.make(static_params={\"g\": jnp.array(9.81), \"l\": jnp.array(1.0),  \"m\": jnp.array(1.0)}),EnvironmentRegistry.PENDULUM.make(static_params={\"g\": jnp.array(9.81), \"l\": jnp.array(1.0),  \"m\": jnp.array(1.0)})]\n",
+    "batched_envs = jax.tree.map(lambda *args: jnp.stack(args), *envs)\n",
```
Comment (hvater) [inline examples/env_functions_examples.ipynb]: I understand that `.make` has the functionality to directly create batched envs, but then you only get the same parameters for all envs.

Probably something like this would be nice to have somewhere, so that users dont need to think about it:

```
def batch_envs(envs: list[CoreEnvironments]) -> CoreEnvironment:
    return jax.tree.map(lambda *args: jnp.stack(args), *envs)
```

File: exciting_environments/acrobot/acrobot_env.py [hunk 14]
```diff
@@ -90,36 +101,34 @@ def __init__(
...
 
         if not control_state:
             control_state = []
 
+        logic = soft_constraints if soft_constraints else acrobot_soft_constraints
+        self.soft_constraints_logic = logic
         self.control_state = control_state
-        self.soft_constraints = soft_constraints
 
         physical_normalizations = self.PhysicalState(**physical_normalizations)
```
Comment (hvater) [inline exciting_environments/acrobot/acrobot_env.py:133]: Bit of a side-topic:

Have you considered what happens if your inputs are scalars/floats instead of arrays?

I think we should cast all `float` in the input dicts to `jnp.array` here. Wdyt?

File: exciting_environments/fluid_tank/fluid_tank_env.py [hunk 15]
```diff
@@ -5,24 +5,36 @@
...
 from dataclasses import fields
 
 from exciting_environments import CoreEnvironment
 from exciting_environments.utils import MinMaxNormalization
 
 
+def fluidtank_soft_constraints(instance, state, action_norm):
+    state_norm = instance.normalize(state)
+    physical_state_norm = state_norm.physical_state
+    phys_soft_const = jax.tree.map(lambda _: jnp.nan, physical_state_norm)
```
Comment (hvater) [inline exciting_environments/fluid_tank/fluid_tank_env.py:21]: Does this mean that there are no constraints for the state in the fluid tank?

While there is no need for a lower constraints as the fluid level cannot go below 0m, there should be an upper constraint at +1. 

I would prefer a constraint -+1 over no constraints.

**Review by hvater — COMMENTED** [review]
Overall I really like the changes. Nice work!

Please look at the comments and implement the suggestions. If you have any questions/differing opinions, I am open for discussion.

File: exciting_environments/core_env.py (hunk 1 above)
Comment (OliverSchw) [reply exciting_environments/core_env.py:18]: I think we need to keep that for now. The parameter tau impacts the structure of certain functions, particularly regarding simulate-ahead logic. Since different values of tau lead to varying output shapes, it cannot be handled as a traceable jax.Array and also should not be varied accross batched envs. Same goes for the solver.

File: exciting_environments/pmsm/pmsm_env.py (hunk 9 above)
Comment (OliverSchw) [reply exciting_environments/pmsm/pmsm_env.py]: Following the same logic as for tau and the solver, deadtime and saturation must be set to static, as they also influence the function's structure and output shape.

**Review by OliverSchw — APPROVED** [review]
I have revised the points raised and also changed the LUT interpolator structure.
